vmdb/app/models/ems_refresh/parsers/foreman.rb

```diff
@@ -0,0 +1,119 @@
+        add_ids(ids, result[:customization_scripts])
+        result[:operating_system_flavors] = operating_system_flavors_inv_to_hashes(inv[:operating_systems], ids)
+        result
+      end
+
+      def configuration_inv_to_hashes(inv, provisioning_manager)
+        result = {}
+        ids = fetch_provisioning_manager_ids(provisioning_manager)
+        result[:configuration_profiles] = configuration_profile_inv_to_hashes(inv[:hostgroups], ids)
+        add_ids(ids, result[:configuration_profiles])
+        result[:configured_systems] = configured_system_inv_to_hashes(inv[:hosts], ids)
+        result[:missing_key] = true if missing_key
+        result
+      end
+
+      def media_inv_to_hashes(media)
+        media.collect do |m|
+          {
+            "manager_ref" => "medium:#{m["id"]}",
+            "type"        => "CustomizationScriptMedium",
+            "name"        => m["name"]
+          }
+        end
+      end
+
+      def ptables_inv_to_hashes(ptables)
+        ptables.collect do |m|
+          {
+            "manager_ref" => "ptable:#{m["id"]}",
+            "type"        => "CustomizationScriptPtable",
+            "name"        => m["name"]
+          }
+        end
+      end
+
+      def customization_scripts_inv_to_hash(media, ptables)
+        media_inv_to_hashes(media) + ptables_inv_to_hashes(ptables)
+      end
+
+      def operating_system_flavors_inv_to_hashes(flavors_inv, ids)
+        flavors_inv.collect do |os|
+          {
+            "manager_ref"           => "operating_system:#{os["id"]}",
+            "name"                  => os["fullname"],
+            "description"           => os["description"],
+            "customization_scripts" => ids_lookup(ids, os["media"], "medium") + ids_lookup(ids, os["ptables"], "ptable")
+          }
+        end
+      end
+
+      def configuration_profile_inv_to_hashes(recs, ids)
+        recs.collect do |profile|
+          {
+            "manager_ref"                    => "hostgroup:#{profile["id"]}",
+            "name"                           => profile["name"],
+            "description"                    => profile["title"],
+            "operating_system_flavor_id"     => id_lookup(ids, profile, "operating_system", "operatingsystem_id"),
+            "customization_script_ptable_id" => id_lookup(ids, profile, "ptable"),
+            "customization_script_medium_id" => id_lookup(ids, profile, "medium"),
+          }
+        end
+      end
+
+      def configured_system_inv_to_hashes(recs, ids)
+        recs.collect do |cs|
+          {
+            "manager_ref"                => "host:#{cs["id"]}",
+            "hostname"                   => cs["name"],
+            "configuration_profile"      => id_lookup(ids, cs, "hostgroup"),
+            "operating_system_flavor_id" => id_lookup(ids, cs, "operating_system", "operatingsystem_id"),
+          }
+        end
+      end
+
+      private
+
+      def fetch_provisioning_manager_ids(manager)
```

Comment:
@kbrock I feel like this is saver logic, do you?  Just going based on the name, I would expect the parser to parse the raw data into a structure that the saver will accept.  Then I would expect the saver to deal with database interaction (id's, relationships, etc.).  If that was the case, the parser could live in the gem as you originally wanted, right?  @Fryguy what do you think?